jcl/src/java.base/share/classes/java/lang/System.java

```diff
@@ -149,7 +149,11 @@
 		platformEncoding = getSysPropBeforePropertiesInitialized(sysPropID_PlatformEncoding);;
 		String definedFileEncoding = getSysPropBeforePropertiesInitialized(sysPropID_FileEncoding);
 		String definedOSEncoding = getSysPropBeforePropertiesInitialized(sysPropID_OSEncoding);
-		if (definedFileEncoding != null) {
+		if (definedFileEncoding != null
+/*[IF JAVA_SPEC_VERSION >= 18]*/
+			&& !definedFileEncoding.equals("COMPAT")
```

Comment:
Can this case happen? The change in system.c shouldn't allow it. Seems this can be fixed natively or in java, but doesn't need changes in both places.

Reply:
Agreed, it is sufficient w/ the change within `getSysPropBeforePropertiesInitialized` which is required to set the property at VM early boot stage.

Reply:
>  the change within getSysPropBeforePropertiesInitialized which is required to set the property at VM early boot stage

I'm curious why the change must be done in the native code, can you pls add a comment in the code to explain it.

Reply:
Added a comment
```
		/*[IF JAVA_SPEC_VERSION >= 18]
		-Dfile.encoding=COMPAT is mapped to platform encoding within getSysPropBeforePropertiesInitialized.
		During early VM boot stage, JITHelpers hasn't been initialized yet, COMPACT_STRINGS related APIs
		might not work properly.
		/*[ENDIF] JAVA_SPEC_VERSION >= 18 */
```